wasi/wasi.c

```diff
@@ -1104,8 +1108,17 @@ wasiFDReaddir(
         inode = entry->d_ino;
         name = entry->d_name;
         nameLength = strlen(name);
-        fileType = wasiFileTypeFromDirentFileType(entry->d_type);
+#if defined (DTTOIF)
+        fileType = wasiFileTypeFromMode(DTTOIF(entry->d_type));
+#else
+        if (lstat(descriptor.path, &entry_stat)) {
+            WASI_TRACE(("fd_readdir: lstat failed: %s", strerror(errno)));
+            return wasiErrno();
+        }
 
+        fileType = wasiFileTypeFromMode(entry_stat.st_mode);
+
+#endif
```

Comment:
Nice, this handles the case where d_type is not available in a more portable way 👍 

We should also handle the case where d_type is available, but is not supported by the file system, maybe like so: 

```suggestion
        // Some operating systems do not provide d_type (Linux, macOS, and some BSDs do).
        // When available, d_type is not supported by all filesystem types.
        // Fall back by using lstat.

#if defined (DTTOIF)
        fileType = wasiFileTypeFromMode(DTTOIF(entry->d_type));
#else
        fileType = wasiFileTypeUnknown;
#endif

        if (fileType == wasiFileTypeUnknown) {
            struct stat entryStat;

            if (lstat(descriptor.path, &entryStat)) {
                WASI_TRACE(("fd_readdir: lstat failed: %s", strerror(errno)));
                return wasiErrno();
            }

            fileType = wasiFileTypeFromMode(entryStat.st_mode);
        }
```

Reply:
I see that `fileType` is defined at the beginning of the function as wasiFileTypeUnknown so no need to redefine it if the macro is not available.  Also, it is not redefined until this point anyway.  Futhermore, the double slash comment type was added in C99 and C++, not present in C89.  The rest looks ok.  I'll commit this with the changes I just mentioned.

Reply:
It is defined at the beginning of the function, but re-assigned in the while loop, so should be reset in each iteration (call of `readdir`).

Good catch on the C++ comment syntax.

Please use the syntax as suggested (spacing between operators, camelCase for variable name